### ISSUE TITLE
Move `addon-info` into addons panel

### ISSUE THREAD

**Issue by shilman** [storybookjs/storybook#1147]
## Why

Currently `addon-info` shows component info in the preview pane alongside each story. This behavior differs from typical addons, that display their info in the addons panel.

Modifying the behavior to use the addons panel would make the addon available for react-native. It also means that addon-info can benefit from addon panel UI improvements in future versions of `storybook-ui`.

EDIT: It would also make the addon compatible with `storyshots`

## What
- [ ] modify the info addon to use the addons panel
- [ ] update docs

**Comment by mthuret** [comment]
Maybe it would be better to split the `addon-info` features to be displayed in different panel? One for the description, one for the source, one for the propTypes? What do you think?

**Comment by Gongreg** [comment]
If you are thinking about splitting it into separate panels it starts to sound like separate addons which already exist.
There are propTypes addon, description addon, source addon.

**Comment by bigassdragon** [comment]
@shilman is this a new addon?

**Comment by shilman** [comment]
@bigassdragon i'm proposing modifying the existing addon, but it could also be a separate addon. it seems like there are a bunch of addons that do similar things in this space, so it would be nice (i think!?) to consolidate and remove duplication of work...

**Comment by bigassdragon** [comment]
@shilman is there a link to the existing addon we are talking about?

I tried looking for it but I have never seen or heard of `addon-panel`, I would not mind having any kind of merge just need to know what it would look like and if we can make and even more awesome addon that could incorporate a lot of slick things to make documentation better.

**Comment by shilman** [comment]
@bigassdragon sorry for the confusion. hopefully this clarifies?

<img width="1143" alt="storybook" src="https://cloud.githubusercontent.com/assets/488689/26608016/7757b74e-45ed-11e7-9687-d89abf814a94.png">

**Comment by bigassdragon** [comment]
@shilman OOOOHHHHHHH. I definitely get it now. yeah I use those.

However I thought there was an addon that already does a panel view of the documentation.

I think you might be trying to do something like this: https://github.com/mihalik/storybook-addon-markdown

However the difference is it would use the `addon-info` plugin so you do not need to create a bunch of markdown files.

Is that correct?

**Comment by shilman** [comment]
@bigassdragon AFAIK:
- `addon-info` looks at proptypes etc. to create an API description semi-automatically
- `addon-markdown` accepts generalized markdown?

I think this should all be simpler, cleaner, and compatible with the rest of the plugins.

**Comment by bigassdragon** [comment]
@shilman correct, I was just making the comparison that you want the docs to show up in one of the panels in your picture above.

**Comment by bigassdragon** [comment]
@shilman I think this idea would be awesome. I kind of thought it would be cool to have the option to have it in the panel and if we put it in the panel let's provide the user the chance to remove the blue button that shows up where you switch between documentation and example.

I think it would be really slick if we can have the ability to show the docs in a panel and also the configuration to hide the blue button because that could be redundant to show.

**Comment by Gongreg** [comment]
@shilman:

Right now there are:

Plugin for documentation: https://github.com/Gongreg/react-storybook-addon-docgen
Plugin for usage: https://github.com/Kilix/storybook-addon-jsx , https://github.com/Gongreg/storybook-usage

As far as I understand the info addon is kinda like a wrapper for these two.

I think it would be good to have simply two separate addons inside monorepo, so users could choose to op out from using one of the addons if they want to.

And then the info addon wouldn't be necessary at all.

**Comment by shilman** [comment]
Thanks @Gongreg that makes a lot of sense. I think part of the problem is that `addon-info` came first and is mentioned all over the place in the docs, so lots of people use it and don't necessarily even know about the replacements. It's a little messy/confusing right now, but I'm sure we can come up with a clean solution here.

**Comment by shilman** [comment]
Here's another reason this should happen: 

https://github.com/storybooks/storybook-addon-a11y/issues/11

**Comment by ranneyd** [comment]
+1 really wish it worked this way

**Comment by shilman** [comment]
Furthermore, we should get rid of the `addWithInfo(title, description, story)` API because it makes it impossible for addons to interoperate. Instead, we can use the pattern used in other addons (e.g. `notes`). 

Either:

```js
storiesOf('MyComponent', module)
  .addDecorator(withInfo(title, description))
  .add('foo', () => <MyComponent />)
  ...
```

Or something like:

```js
storiesOf('MyComponent', module)
  .add('foo', () => <WithInfo title='my title' description='my desc'><MyComponent /></WithInfo>)
```

**Comment by thearegee** [comment]
+1 for the first option

**Comment by tmeasday** [comment]
@shilman yes this seems like a very easy thing to change and a good practice to establish. A lot of addons out there follow the `addWithX` approach probably because they are just imitating the canonical info addon.

**Comment by tmeasday** [comment]
I looked briefly into replacing `addon-info` with the docgen and jsx addons that @Gongreg [mentioned](https://github.com/storybooks/storybook/issues/1147#issuecomment-305114066) above. The issues I ran into (which are probably why he forked them) are:

1. The docgen plugin doesn't seem to have been updated for 3.0 yet, and doesn't quite work.
2. The JSX plugin requires the use of `.addWithJSX()` (strictly speaking this isn't a step backwards). Still it would be good to allow the decorator pattern @shilman mentioned above.

However these issues are probably easily solved. I am happy to help with updating these packages and bringing them into core alongside deprecating the info addon if the respective authors would like that to happen (@wcastand, @mihalik?)

**Comment by wcastand** [comment]
Hi, I wouldn't mind putting jsx into the monorepo but i'm not okay with the `WithNotes` pattern.
Addon for panels shouldn't be in my stories.
We know this is a issue right now and we discussing on how to change the addon API to allow this kind of patterns and avoid the overwrite of `.add()`

**Comment by tmeasday** [comment]
@wcastand -- do you mean you don't like the fact that the story gets "polluted" with the decorator?

Is there a specific reason you don't like it (i.e. does it stop you doing something?), or is it more a matter of style? (I don't really like it either)

I am looking at this from a short-term perspective. The issue with `.addWithX()` is you can only use one addon that uses it! This is a problem for adoption of addons.

Also the nice thing about the decorator pattern is you can apply it to a whole chapter or even all stories in a natural way using an API that users are familiar with.

My slightly longer term suggestion is to solve #993 and pass options in the `.add()` call. This avoids the decorator problem.

A new addons API would be good! But it seems a bit further off.

**Comment by tmeasday** [comment]
(It is of course possible to offer both APIs from the package)

**Comment by Gongreg** [comment]
@tmeasday You are totally right, I forked them because they were not updated.

@wcastand Why can't you reach storyFn() while using addDecorator() pattern?

The addon-usage plugin works this way.

The problem with addDecorator is that the order is important.
That allows to use intersting combinations of addons, but problem is that addons pollute storyfn().

So if user adds addDecorator(WithJSX) as a first addon, you will get same storyFN() as using addWithJsx().

**Comment by wcastand** [comment]
It's mostly a matter of style. separation of my stories and storybook's helpers(addons, etc)
I don't mind decorators at all, i use them for thing like Redux, Fela. But decorators are for app related needs.

In my opinion, they shouldn't be used for storybook's panels.
I won't change the api of the addon JSX for a temporary fix that breaks the plugin for older version.

For solutions, like i said, i made a proposition with decorators and middleware.

@Gongreg i can access storyFn from decorators but storybook's addons are not their use in my opinion. It's what confuse people about addon in storybook. They shouldn't be in the stories at all.

**Comment by tmeasday** [comment]
> I won't change the api of the addon JSX for a temporary fix that breaks the plugin for older version.

Well, like I said it is possible to offer both APIs, so it's not a breaking change.

**Comment by Gongreg** [comment]
@wcastand , so ok, we can both agree that there are problems with current addons implementation.
But having to explicitly tell in every story that you have to use your addon (addWithJsx) while limiting other addon usage is not a solution too :) Also your api doesn't allow me to say "use this addon in all the stories and I don't care about it at all anymore".


At least having as an option, to add it as a decorator would be a nice step forward. 
Then you could definitely add it to monorepo, since it would be compactible with other plugins.

**Comment by wcastand** [comment]
English is not my native langage so i'm sorry if i sound "bad", it's not the case and i don't mind discussing it (just in case :)).

@tmeasday right now, i don't think it's a good idea but if someone want to make that happen and do a PR i don't mind helping him and publish it :)

@Gongreg It's not a solution right now. That's why i started to bring the need to rework the addon API in the first place.
the API proposal that i made allow storybook wide decorators and middleware (so you can configure a redux decorator and a JSX middleware for all of your stories) [Draft](https://gist.github.com/wcastand/caf232a2d1d16b4b08ec17e38a9ecbdf#file-ADDON-Example-js)

Why i think it's a bad idea to make my addon and all the addon using the decorator pattern as a temporary fix for storybook ? 
Because the temporary fix will probably become the permanent solution if everything work even if it means blending decorator specific to your app et decorator specific to storybook panels(middleware).

I prefer to have a none perfect solution for a while and work on a rework so we go on good basis instead of making something work right now and never change it because it works.

it's only my opinion and i get it if people doesn't agree but i work like this:
- make it work
- make it right
- make it fast

For me storybook is 'make it work' (it works really well :D). Putting storybook in a monorepo, using a proper namespace is a way forward the make it right and a rewrite ( at least partial ) is for me the next step.

The rework of the Addon API is needed (in my opinion) because the addon API doesn't fit the use and like everyone here said : it's not clear and well defined what is a decorator or how do i make a storybook panels without polluting my stories or overwrite`.add` (not possible if you want to access storyFn).

I think we can agree on the api and make it quite quickly, it probably won't be perfect but we can iterate.

In short, if someone wants to make JSX works with decorators i'll help him and i'll add it to the package as a 'temporary solution' but i won't do it myself because i think it's not a good idea until we figure it out the new addon API.

i'm on storybook's slack if you want to discuss with me and avoid pollute this issue with not so related content.

**Comment by tmeasday** [comment]
I'm happy to contribute a PR. I'll send it through in the next couple of days.

I think generally I agree with you @wcastand except on one point; I would argue that `.addWithX` doesn't "work" (IMO) and we need to resolve that problem before we make it right :)

**Comment by lucasmotta** [comment]
What about having an official `addWith` API, that allows you to compose different addons together:

## Single Addon
```js
story.addWith(info('Description of this component for the info addon'))(
  () => <Component />
)
```

## Composes Multiple Addons
```js
story.addWith(
  info('Description of this component for the info addon'),
  jsx('This is for the JSX addon')
)(() =>
  <Component />
)
```

Or even having that logic implemented on the existing `add` method.

**Comment by tmeasday** [comment]
@lucasmotta it's interesting because this is very nearly possible with the current API, you could have

```js
story.add('name',
  info('Description of this component for the info addon')(
    jsx('This is for the JSX addon')(
      () => <Component />)));
```

In the current API. The only thing I would [propose](https://gist.github.com/tmeasday/775c5572271121bb2a71153ac5f2cd7a) is to enable moving the decorators together (which ends up being similar to your idea):

```js
story.add('name', () => <Component />), {
  decorators: [
    info('Description of this component for the info addon'),
    jsx('This is for the JSX addon'),
  ],
});
```

(This way treats the decorators a little less specially and allows room to grow and add other options later).

**Comment by IanVS** [comment]
This is an old issue, and I'm not sure exactly where it stands or if it's still planned to be changed.  But as a new storybook user, I have to say that I find the current behavior very useful.  I am implementing storybook as a kind of "styleguide", and this addon allows me to easily create a "page" for each component with some vital information like a generic description, example code, and prop types.  

I only bring this up to provide a voice in support of keeping this in the preview area.  I love being able to tweak knobs and see how the example component updates, while also seeing the rest of the info that this add-on provides (without having to toggle between add-on tabs).

**Comment by waynebrantley** [comment]
Perhaps it should be optionally in a tab.
As it is, if you use storyshot with or without images, this addon has to be disabled.
If not, any change to your docs causes a 'change' as the snapshot has all the info in it.

**Comment by charrondev** [comment]
Has there been any additional movement on this? I'd be interested in helping out, but I'd want to get some clear acceptance criteria before sinking time into it.

**Comment by hipstersmoothie** [comment]
Definitely needs some docs for using the pure docgen plugin. The only one i could find for typescript I could not get to work. Whereas the info plugin works fine

**Comment by hipstersmoothie** [comment]
I've published `storybook-addon-react-docgen`. It uses the docgen information just like the info addon but places the information in a `Props` tab in the the addons panel. It works quite well with the notes notes-addon 

https://www.npmjs.com/package/storybook-addon-react-docgen

**Comment by JamesIves** [comment]
@shilman Is there plans to implement this? I'm running into this problem where the addon info button is blocking parts of my component preview and would love for this data to appear in the addons panel.

<img width="595" alt="Screen Shot 2019-03-22 at 2 01 10 PM" src="https://user-images.githubusercontent.com/10888441/54843676-57f05e80-4cab-11e9-93ea-a74f678cb602.png">

**Comment by shilman** [comment]
@jamesives Yeah it's been on the back burner for a long time but I have a prototype running and plan to release it in 5.1. we're putting a focus on component documentation now that 5.0 has been successfully released

**Comment by JamesIves** [comment]
Excellent! Is there an ETA for that release?

**Comment by shilman** [comment]
Hopefully we can get that feature  available in  prerelease by mid-April and a solid first cut by May

**Comment by avrilpearl** [comment]
@JamesIves --> you can define this in the configuration of the info addon as well: e.g:

```
addDecorator(
    withInfo({
      styles: {
        button: {
          base: {
            fontFamily: 'sans-serif',
            fontSize: '14px',
            fontWeight: '500',
            display: 'block',
            position: 'fixed',
            border: 'none',
            background: '#14558f',
            color: '#fff',
            padding: '5px 15px',
            cursor: 'pointer',
          },
          topRight: {
            bottom: 0,
            right: 0,
            top: 'unset',
            borderRadius: '5px 0 0 0',
          }
        }
      }
    })
  );
```

**Comment by JamesIves** [comment]
Thanks @avrilpearl ! This is a great workaround.

**Comment by inspiratweb** [comment]
Do we have any news about this @shilman?

**Comment by shilman** [comment]
@inspiratweb Addon-docs has been in alpha for over a month and is nearing beta. It solves a ton of problems with addon-info and adds a bunch of new features as well.

Read about it here: https://medium.com/storybookjs/storybook-docs-sneak-peak-5be78445094a

And get started here: https://docs.google.com/document/d/1un6YX7xDKEKl5-MVb-egnOYN8dynb5Hf7mq0hipk8JE/edit?usp=sharing

**Comment by shilman** [comment]
FYI `addon-info` is being superceded by `addon-docs`, which fixes a bunch of bugs and is easier to maintain. It’s reached release candidate (RC) status and will be properly released soon. Please give it a try! https://medium.com/storybookjs/storybook-docspage-e185bc3622bf

@kevin940726 @kohlmannj @zillding @ajwhite @ranneyd @dslounge @lucasmotta @seanking11 @jeffmicklos @igor-dv @evenchange4 @DimaRGB @levithomason @saponifi3d @L0stSoul @clementallen @louisfly @mariosanchez @tourshi @charlieclark @jalooc @nickarora @vincenthirtz @lb- @denkristoffer @mrjackdavis @zolk @patrick-gill @bmson @mindtricksdev @ruxiang05 @joshsylvester @christiannaths @keyboard-clacker @feyssericAzalead @Gidgidonihah @lslima91 @sapkra @kangax @wyattdanger @andriichumak @kylemh @ewolfe @svict4 @gautier-lefebvre @jack-sf @MihaChicken @charrondev @bradfrost @motanelu @aurerua @jmarceli @kngroo @Leko @sellsjonathan @iannbing @BenRichter @snnsnn @leoyli @Lehakos @Jimmerz28 @josephfusco @JeffersonCarvalh0 @airtonix @JasonTheAdams @JamesIves @UlanaXY @d4rekanguok @zachguo @fnhipster @maddesigns @techieshark @gwyneth-bassett @bryancuster @MarcoDaniels @DesignByOnyx @jakubriedl @inspiratweb @hsz @kaitlynhova @mthuret @will-stone @kroeder @polem @char0n @Gongreg @bigassdragon @jonyeezs @wcastand @bolonio @brunoreis @thearegee @tomhuze @tmeasday @nutgaard @UsulPro @MrOrz @otorrillas @josephdonnelly @IanVS @waynebrantley @ndelangen @strothj @Geril @hipstersmoothie @chrisvxd @3ddmoo @gehtman @storywr @Pickra @drewdecarme @CPedrini @petermooibroek @crutchcorn @binomialstew @avrilpearl @edmulraney @dmahnkopf @jpvincent @haldunanil

**Comment by shilman** [comment]
This is the most up-voted issue in the Storybook repo and I'm closing it. If anybody wants to pick this up, feel free to create a PR.

For those who care, here's the full explanation:
- `addon-docs` is the successor to `addon-info`. For the past few months, I've been redirecting all `addon-info`-related issues to `addon-docs`.
- `addon-docs` contains a superset of `addon-info`'s features, and also contains many bugfixes.
- `addon-docs` **should** be better than `addon-info` in every way. If you disagree, please feel to file an issue, and we'll do our best to fix it. It should be an attractive, painless upgrade, and if it's not it means that we're doing something wrong.

Thanks for using `addon-info`; please upgrade to `addon-docs` today!

**Comment by DiegoBM** [comment]
It would be wise to redirect from the _Info_ addon to the _Docs_ addon in the _Info_ page on Github, or warn about the deprecation in some sort of way there, just so that people know that the _Info_ addon is not being supported anymore and that they should use _Docs_ instead, and avoid wasting time trying to make it work.

FYI, installing the _Info_ addon breaks storybook (it removes react from node_modules)

**Comment by shilman** [comment]
@DiegoBM done. thanks!
